metadata/metadata_test.go

```diff
@@ -54,3 +60,246 @@ func TestValidateHeaderInfo(t *testing.T) {
 		assert.Equal(t, e, tt.err, "failing test: %v (%v)", idx, tt)
 	}
 }
+
+func TestValidateTypeInfo(t *testing.T) {
+	var validateTests = []struct {
+		in  MetadataTypeInfo
+		err error
+	}{
+		{MetadataTypeInfo{}, ErrInvalidTypeInfo},
+		{MetadataTypeInfo{Rootfs: ""}, ErrInvalidTypeInfo},
+		{MetadataTypeInfo{Rootfs: "image-type"}, nil},
+	}
+
+	for _, tt := range validateTests {
+		e := tt.in.Validate()
+		assert.Equal(t, e, tt.err)
+	}
+}
+
+func TestValidateMetadata(t *testing.T) {
+	var validateTests = []struct {
+		in  Metadata
+		err error
+	}{
+		{Metadata{}, ErrInvalidMetadata},
+		{Metadata{make(map[string]interface{})}, ErrInvalidMetadata},
+		{Metadata{map[string]interface{}{}}, ErrInvalidMetadata},
+		{Metadata{map[string]interface{}{"": nil}}, ErrInvalidMetadata},
+		{Metadata{map[string]interface{}{"key": nil}}, ErrInvalidMetadata},
+		{Metadata{map[string]interface{}{"key": "val"}}, ErrInvalidMetadata},
+		{Metadata{map[string]interface{}{"DeviceType": "type"}}, ErrInvalidMetadata},
+		{Metadata{map[string]interface{}{"DeviceType": nil, "ImageID": "image"}}, ErrInvalidMetadata},
+		{Metadata{map[string]interface{}{"DeviceType": "device", "ImageID": "image"}}, nil},
+		{Metadata{map[string]interface{}{"DeviceType": "device", "ImageID": "image", "Data": "data"}}, nil},
+	}
+
+	for _, tt := range validateTests {
+		e := tt.in.Validate()
+		assert.Equal(t, e, tt.err, "failing test: %v", tt)
+	}
+}
+
+func TestValidateFiles(t *testing.T) {
+	var validateTests = []struct {
+		in  MetadataFiles
+		err error
+	}{
+		{MetadataFiles{}, ErrInvalidFilesInfo},
+		{MetadataFiles{Files: []MetadataFile{}}, ErrInvalidFilesInfo},
+		{MetadataFiles{Files: []MetadataFile{{File: ""}}}, ErrInvalidFilesInfo},
+		{MetadataFiles{Files: []MetadataFile{{File: "file"}}}, nil},
+		{MetadataFiles{Files: []MetadataFile{{File: "file"}, {}}}, ErrInvalidFilesInfo},
+		{MetadataFiles{Files: []MetadataFile{{File: "file"}, {File: "file_next"}}}, nil},
+	}
+	for idx, tt := range validateTests {
+		e := tt.in.Validate()
+		assert.Equal(t, e, tt.err, "failing test: %v (%v)", idx, tt)
+	}
+}
+
+func MakeFakeUpdateDir(updateDir string, elements []MetadataDirEntry) error {
+	for _, elem := range elements {
+		if elem.IsDir {
+			if err := os.MkdirAll(path.Join(updateDir, elem.Path), os.ModeDir|os.ModePerm); err != nil {
+				return err
+			}
+		} else {
+			if _, err := os.Create(path.Join(updateDir, elem.Path)); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+var dirStructOK = []MetadataDirEntry{
+	{Path: "files", IsDir: false},
+	{Path: "type-info", IsDir: false},
+	{Path: "meta-data", IsDir: false},
+	{Path: "checksums", IsDir: true},
+	{Path: "checksums/image.sha", IsDir: false},
+	{Path: "signatures", IsDir: true},
+	{Path: "signatures/iamge.sig", IsDir: false},
```

Comment:
You've misspelled `image` as `iamge` in all the test cases. This would actually be an error, so you could consider testing for this. All the checksums and signatures must have a 1-to-1 name relationship with all the payload files.